pandas/core/generic.py

```diff
@@ -4429,14 +4430,18 @@ def astype(self, dtype, copy=True, errors='raise', **kwargs):
                 if col_name not in self:
                     raise KeyError('Only a column name can be used for the '
                                    'key in a dtype mappings argument.')
-            from pandas import concat
             results = []
             for col_name, col in self.iteritems():
                 if col_name in dtype:
                     results.append(col.astype(dtype[col_name], copy=copy))
                 else:
                     results.append(results.append(col.copy() if copy else col))
-            return concat(results, axis=1, copy=False)
+            return pd.concat(results, axis=1, copy=False)
+
+        elif is_categorical_dtype(dtype) and self.ndim > 1:
+            # GH 18099: columnwise conversion to categorical
+            results = (self[col].astype(dtype, copy=copy) for col in self)
+            return pd.concat(results, axis=1, copy=False)
```

Comment:
I don't think using concat is the best way here. For example, this will loose information about the columns object (its type, metadata such as its name). 
Can't we in a loop assign it back into the frame? (or is that less efficient?)

Reply:
this is idiomatic and how things are handled elsewhere. 

> For example, this will loose information about the columns object (its type, metadata such as its name)

it will preserve as these are all the *same* index. of course if you have a counter-example

Reply:
I am speaking about the *columns*, not index. So just take an example with CategoricalIndex, RangeIndex, .. or one with a name, and the resulting columns will not be the same.

Reply:
hmm, we do this else where in astype. So this is obviously not tested. ok In any event, using concat is vastly more performant. I suppose you just set the columns after.

Reply:
I've opened #19920 to address the (an?) existing occurrence of this bug.  A quick check makes it look like @jreback's suggestion of setting the columns afterwards should fix things in both places, e.g.
```python
result = pd.concat(results, axis=1, copy=False)
result.columns = self.columns
return result
````
Since this is relatively small, I'm planning to make the fix/add tests in this PR.  Can address it in a separate PR if that would be preferable though.

Reply:
Both (here or in other PR) are fine for me. However, if we also want to fix this for the other cases (not only for categorical), it might be cleaner as a separate PR.

Reply:
yes this is fine, I see you opened another one which is good.